### ISSUE TITLE
forgetting to use `instance` should be a type error

### ISSUE THREAD

**Issue by safareli** [NiGhTTraX/strong-mock#231]
Not quite sure how, but would be nice if forgetting to call `instance` was a type error.

As a last resort linter rule might be do the trick as well, but TS solution would be best.

(btw getting message that "hey you might forgot to call `instance`" was super helpful 👍 )

Thanks for the great lib!

**Comment by NiGhTTraX** [comment]
Hey @safareli, yeah, I'd love to have that as a type error, but I haven't found any good solution that both covers all the cases and that preserves a good IDE experience. `Mock<T>` needs to be compatible with the call signature of  `T` so you can call it inside `when()` as you would call it in production code, and transforming `T` in any way can cause IDEs to not update your tests when you e.g. rename a method.

To get a better sense this, imagine you want to mock `() => void`. I can't think of any type that preserves the call signature, but is ultimately incompatible. If anyone has any ideas I'd love to hear them 👂 

---

P.S.: I'm glad the error message helped 😄  It only triggers on the 2nd call to a mock without `instance`, because the first one creates an expectation that is not finished when the second one happens.

**Comment by parisholley** [comment]
thought.. when you call `when()`, it could set a global (eg: `when.start = true`), when doing `when(mock.expect())`, the mock proxy will check if when.start is set, otherwise throw an exception. then once `when()` returns, it would set it back to false. so long as `when()` isn't pushing anything on the event loop, you shouldn't have any race conditions

**Comment by parisholley** [comment]
for a type solution, perhaps `mock()` could actually adjust all the return types to look like `() => T` or `() => void`. that way, `when()` could still infer the response type for `thenResolve`/`thenReturn`, etc, and most consumers of the mock would see errors if using directly, since the return type isn't `T` but rather `() => T`, and calling `instance(mock)` would strip the function returns and change the signatures to `T` or `void`.

that idea is hypothetical of course :p, may not be 100% possible. only edge case the above wouldn't cover, is `() => void`, as if you aren't asserting the result, there really isn't a way to see a type error (though you may be able to get a linter involved with something hacky like `() => Promise<void>`, that way it will complain if you aren't handling promises.

**Comment by safareli** [comment]
With `when(mock.expect())` as js is not lazy language `mock.expect()` will be evaluated first. so even if when set's `when.start` to true for `mock.expect()` it would be falkse.

About the type solution instead of `() => __` wrapping you can have some other wrapping but then I think you _**might**_ loose the renaming/refactoring capabilities

**Comment by NiGhTTraX** [comment]
Like @safareli said, the first solution won't work because of evaluation order.

The second solution is the best bet and might be a reasonable tradeoff, but it needs to be tested thoroughly in IDEs since a mapped type + a conditional type (to handle functions vs objects/classes) might lose track of the original type in refactorings.

**Comment by NiGhTTraX** [comment]
Small update: I tried the solution discussed here to deconstruct `T` from `mock<T>` and then reconstruct it in `instance<T>`, and unfortunately the type checker can't connect the dots between the 2. Here's what I tried, in case someone has any suggestions:

```ts
export type MockReturn<T> = { r: T };

type ObjectMock<T> = {
  [K in keyof T]: T[K] extends (...args: infer A) => infer R
    ? (...args: A) => MockReturn<R>
    : MockReturn<T[K]>;
};

export type Mock<T> = T extends (...args: infer A) => infer R
  ? ((...args: A) => MockReturn<R>) & ObjectMock<T>
  : ObjectMock<T>;
```

With the above, `instance = <T>(mock: Mock<T>) => T` doesn't work correctly, neither does using the reverse of the types.

----

The other option that I'm considering is changing how `when` works — from `when(call)` to `when(() => call)`. This will allow differentiating between a call during an expectation versus one during the test since the call will be lazily run inside `when`. Obviously, this would be a breaking change, and a step away from simplicity, but it would also allow mocking setters by using `when(() => { foo.bar = 'baz'; })`.

----

I will keep this issue open for visibility and for gathering feedback and/or suggestions.

**Comment by parisholley** [comment]
breaking syntax that would still be runtime but fix the lazy issue? `when(mock.expect)(arg1,arg2).thenResolve()` 

@NiGhTTraX this snippet seems to work for me, unless your referring to type issues with strong-mock function implementations (i stubbed when/instance in this code)?

```typescript
export type MockReturn<T> = { r: T };

type ObjectMock<T> = {
  [K in keyof T]: T[K] extends (...args: infer A) => infer R
    ? (...args: A) => MockReturn<R>
    : MockReturn<T[K]>;
};

export type Mock<T> = T extends (...args: infer A) => infer R
  ? ((...args: A) => MockReturn<R>) & ObjectMock<T>
  : ObjectMock<T>;

interface MyInterface {
  foo(): void;
  bar(): string;
  baz(arg: string): string;
  bum: number;
}

function when<T>(call: MockReturn<T>): Stub<T> {
  return null;
}

function instance<T>(mock: Mock<T>): T {
  return null;
}

const mock: Mock<MyInterface> = null;

when(mock.foo()).thenReturn();
when(mock.bar()).thenReturn('string');
when(mock.baz('string1')).thenReturn('string2');
when(mock.bum).thenReturn(5);

const unmocked = instance(mock);
```

**Comment by NiGhTTraX** [comment]
@parisholley it doesn't work correctly for function types, where `instance` infers `unknown`. Moreover, turns out that the combination of mapped types + conditional types [break function overloads](https://www.typescriptlang.org/play?#code/KYDwDg9gTgLgBDAnmYcCyEDGBrASsGAVygDsAeAFQD44BeOAbzigC44K4BfAbgChekKOAHkARgCtgmGBhyUa9BrzhwA2gGk4ASxJxswRBABm7ALpsKG03FAxgJACYBnOAAoAdJ4CGUAOZO2HSNgKDgAQQBKOhogkLhcZRU4AH43T3cffzZI6PQsPAJiclwqRJU2WQKiUkorKj4eflBIWARkVEr5OnYbEDtHFw9vPwDtEmDQnNoY8biElVTXIYyR7KjpvJx8auKqKIAyEQkpGXz5RLYxSWlO6j5eTAAbLycXADEICEZE0R9XCLYThgUB0vj4Kl+UFcIDYJEIAFtRCEAXA4YiQuC4JDoclYQikVAUWiCXAAD5wIEgki+b5JZiFUhwABMAGZMZxeBzeA4pM8oKhMBASED6V5Hh8IGwJXweU8fAKhSL4fkJRUzhL6vx+WKJe5sRE+NrxZ89X9WQb+MqcLr9XwgA) (see also https://github.com/microsoft/TypeScript/issues/29732).

**Comment by NiGhTTraX** [comment]
Summarizing the available options at this point:

#### Option 1 - mapped types

This changes the return type of every property in `T` (and `T` itself, in case it's a function) to something incompatible, while still preserving the original type for access in `when`.

**Pros**: It preserves the syntax of current API, makes nested property access a type error (currently it's a runtime exception).
**Cons**: Brittle and could break unforeseen edge cases in either mock types, or IDE refactoring. If we go with this and run into one of those edge cases, reverting would most likely be the only option to fix it. It also doesn't cover `() => void` types.

#### Option 2 - callback in `when`

This makes the expectation evaluation lazy and gives us the chance to put the proxy in recording mode before the call.

```ts
when(() => foo.bar()).thenReturn('baz')
console.log(foo.bar()); // baz
```

**Pros**: pretty straightforward to implement, enables setter expectations, preserves original type, gets rid of `instance`.

**Cons**: breaking change, extra syntax.

#### Option 3 - extra property in mock

This moves the original type under an extra property on the mock e.g. `mock.expect`. `instance` could be moved to `mock.instance`.

```ts
when(foo.expect.bar()).thenReturn('baz');
console.log(foo.instance.bar()); // baz
```

**Pros**: similar implementation cost to option 2, preserves original type, can get rid of `instance`.

**Cons**: breaking change, extra syntax, doesn't enable setter expectations like option 2, IMO the syntax gets slightly confusing for function types (`when(foo.expect())`).

----

My current favorite is option 2 since it has the most advantages, and I think the migration is easy to understand for users of the current API.

**Comment by parisholley** [comment]
what are your thoughts on something like this:

```typescript
// make the requirement of record a global flag, existing projects could just disable it
record(() => {
  // maintain existing api
  when(mock.foo()).thenReturn();
  when(mock.bar()).thenReturn('string');
  when(mock.baz('string1')).thenReturn('string2');
  when(mock.bum).thenReturn(5);
});

// if mock was not defined in a record(), throw exception on first use?
// eliminates the need for instance()
mock.foo();
```

or alternatively

```typescript
record(mock, () => {
  // maintain existing api
  this.foo().thenReturn();
  this.bar().thenReturn('string');
  this.baz('string1').thenReturn('string2');
  this.bum.thenReturn(5);
});

// if mock.record() was not invoked, throw exception on first use?
mock.foo();
```

**Comment by NiGhTTraX** [comment]
Hmm, not a bad idea @parisholley, it reminds me of [React's `act`](https://reactjs.org/docs/test-utils.html#act). I've updated my summary above to include this option. It's very similar to option 2, but it brings with it an extra import, compared to extra syntax inside `when`.

Just to make sure I understand your proposal fully, what did you mean by this?

> // [...] existing projects could just disable it

How could existing projects work without requiring the new callback?

**Comment by parisholley** [comment]
@NiGhTTraX we could expose a global config like we do for the recent matcher change:

```typescript
setDefaults({
  recording: false
})```

when recording = false, current api behavior is used. you have to use the `instance` method, but you can define mock behavior at any point. using `record()` would throw an exception

when recording = true (as say, a default in a new major release), using `instance()` would throw an exception. interacting with a mock would throw an exception like it does today, if that behavior wasn't recorded.

we effectively just change the mock proxy implementation based on the flag

**Comment by NiGhTTraX** [comment]
@parisholley thanks for clarifying. This made me realize that forgetting to use `record()` with that solution would still be a runtime error, not a type error. For that reason I would rate the solution lower than the other ones.

**Comment by parisholley** [comment]
@NiGhTTraX not sure i follow, options 2.1 and 2.2 both eliminate the need for instance(), so there isn't a `type error` problem to solve?

**Comment by NiGhTTraX** [comment]
Yes, but option 2.1. forces you to pass a callback inside `when` to define your test expectations, and won't compile without it, while option 2.2. needs to wrap the expectations in `record`, and will throw at runtime if you forget.

**Comment by NiGhTTraX** [comment]
Hey @safareli, @parisholley, I just published [8.0.0-beta.0](https://github.com/NiGhTTraX/strong-mock/blob/v8.0.0-beta.0/CHANGELOG.md#800-beta0-2022-07-15) that removes `instance` and adds callbacks inside `when`. I would appreciate it if you could try it out and let me know what you think.

Migration summary:

```diff
-import { instance, mock, when } from 'strong-mock';
+import { mock, when } from 'strong-mock';
 
 const fn = mock<(x: number) => number>();
 
-when(fn(42)).thenReturn(100);
+when(() => fn(42)).thenReturn(100);
 
-console.log(instance(fn)(42));
+console.log(fn(42));
```

**Comment by NiGhTTraX** [comment]
[V8](https://github.com/NiGhTTraX/strong-mock/blob/73ade03f477b6422a74d2faa963aef54d7c7f27a/CHANGELOG.md#800-2022-10-02) is now released!
